### PR TITLE
PM2020-19: Add Product Typology Filter

### PR DESCRIPTION
**Pull request by pierallard**
**Description (for Contributor and Core Developer)**

Adds a new filter in product grid

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | :+1: 
| Added legacy Behats               | :-1: 
| Added acceptance tests            | :-1: 
| Added integration tests           | :-1: 
| Changelog updated                 | :-1: 
| Review and 2 GTM                  | :-1: 
| Micro Demo to the PO (Story only) | :-1: 
| Migration script                  | -
| Tech Doc                          | -

### CODE REVIEW

File: src/Oro/Bundle/PimFilterBundle/Filter/Product/ProductTypologyFilter.php [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    public function apply(FilterDatasourceAdapterInterface $ds, $data)
+    {
+        $data = $this->parseData($data);
+        if (!$data) {
+            return false;
+        }
+
+        switch (current($data['value'])) {
+            case self::PRODUCT_TYPOLOGY_SIMPLE:
+                $this->util->applyFilter($ds, 'family_variant', Operators::IS_EMPTY, null);
```
Comment (nmarniesse) [inline src/Oro/Bundle/PimFilterBundle/Filter/Product/ProductTypologyFilter.php:26]: Should we also filter on product only? This filter removes product models?

**Review by nmarniesse — APPROVED** [review]
Added an integration test should be great IMO (see LabelOrIdentifierFilterIntegration.php for example)

File: src/Oro/Bundle/PimFilterBundle/Filter/Product/ProductTypologyFilter.php (hunk 1 above)
Comment (pierallard) [reply src/Oro/Bundle/PimFilterBundle/Filter/Product/ProductTypologyFilter.php:26]: Having "no family variant" means you only have simple products, not variant nor product models.

Comment (nmarniesse) [reply src/Oro/Bundle/PimFilterBundle/Filter/Product/ProductTypologyFilter.php:26]: :+1:

Comment (nmarniesse) [reply src/Oro/Bundle/PimFilterBundle/Filter/Product/ProductTypologyFilter.php:26]: So having a family variant means variant or product model? It's for the `PRODUCT_TYPOLOGY_VARIANT` case just below

Comment (pierallard) [reply src/Oro/Bundle/PimFilterBundle/Filter/Product/ProductTypologyFilter.php:26]: Exaclty

Comment (nmarniesse) [reply src/Oro/Bundle/PimFilterBundle/Filter/Product/ProductTypologyFilter.php:26]: This is not a problem that the "variant" filter returns variant AND product model? We don't have to filter on product too?

**Review by mmetayer — APPROVED** [review]

File: src/Oro/Bundle/PimFilterBundle/Filter/Product/ProductTypologyFilter.php (hunk 1 above)
Comment (pierallard) [reply src/Oro/Bundle/PimFilterBundle/Filter/Product/ProductTypologyFilter.php:26]: Variant filter returns variant AND product models ; but the "aggregator" removes the variant if we don't have conditions on the current level.
The "filter on product" is the "ungroup/group" feature, and it's another PR.

Comment (nmarniesse) [reply src/Oro/Bundle/PimFilterBundle/Filter/Product/ProductTypologyFilter.php:26]: Ok I understand now. If we filter on "variant", it does not return only variant products :roll_eyes: ... The naming is very confusing, but maybe it's just me. In fact it's a "simple product" filter with true/false value (I'm not totally convinced by this naming too)

Comment (mmetayer) [reply src/Oro/Bundle/PimFilterBundle/Filter/Product/ProductTypologyFilter.php:26]: probably smth we'll discuss tomorrow (ubiquitous language)

**Pull request merged**